extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java

```diff
@@ -76,33 +83,82 @@ public TaskNotRunnableException(String message)
   private final HttpClient httpClient;
   private final ObjectMapper jsonMapper;
   private final TaskInfoProvider taskInfoProvider;
+  private final Duration httpTimeout;
   private final RetryPolicyFactory retryPolicyFactory;
-
-  public KafkaIndexTaskClient(HttpClient httpClient, ObjectMapper jsonMapper, TaskInfoProvider taskInfoProvider)
+  private final ListeningExecutorService executorService;
+  private final long numRetries;
+
+  public KafkaIndexTaskClient(
+      HttpClient httpClient,
+      ObjectMapper jsonMapper,
+      TaskInfoProvider taskInfoProvider,
+      String dataSource,
+      int numThreads,
+      Duration httpTimeout,
+      long numRetries
+  )
   {
     this.httpClient = httpClient;
     this.jsonMapper = jsonMapper;
     this.taskInfoProvider = taskInfoProvider;
+    this.httpTimeout = httpTimeout;
+    this.numRetries = numRetries;
     this.retryPolicyFactory = createRetryPolicyFactory();
+
+    String threadId = String.format("KafkaIndexTaskClient-%s", dataSource);
+    this.executorService = MoreExecutors.listeningDecorator(Execs.multiThreaded(numThreads, threadId + "-%d"));
   }
 
-  public void stop(String id, boolean publish)
+  public void close()
   {
-    submitRequest(id, HttpMethod.POST, "stop", publish ? "publish=true" : null, true);
+    executorService.shutdown();
   }
 
-  public void resume(String id)
+  public Boolean stop(final String id, final boolean publish)
```

Comment:
It gets wrapped in a Callable and ListenableFuture in stopAsync() which is why it can't be a primitive.